### PR TITLE
nixos/printers: declarative configuration

### LINKED ISSUE #52776: Declarative cups configuration
## Issue description
Allow configuration of a cups printer by pointing at a ppd file and specifying a single URL in configuration.nix.

Currently, such configuration requires clicking through a web-interface.

## Technical details
18.09

### PR DESCRIPTION
**Pull request by florianjacob**
###### Motivation for this change
This saves one from going through the cups webinterface each time one sets up a new NixOS machine.
I created a half-stateful solution similar to `mysql.ensureDatabases`, which uses `lpadmin` etc. to create the configured printers and force them in the right configuration, without having to touch the deep parts of cups. Works quite well, at least for me.

Resolves #52776, supersedes #23684.

Tests, comments & proposals for improvements welcome.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-1903.xml
Comment (Mic92) [inline nixos/doc/manual/release-notes/rl-1903.xml]: Printers or databases?

```suggestion
     These options will never delete existing printers and users, especially
```

**Comment by deliciouslytyped** [comment]
A note: In my limited testing with lpadmin, if you add an existing printer again, any printer settings you have changed seem to be reset. (so for example if reconfiguration calls lpadmin again)

Not a breaking thing, but possibly a bit annoying.
What is the behavior of "oneshot"?

File: nixos/doc/manual/release-notes/rl-1903.xml
Comment (florianjacob) [reply nixos/doc/manual/release-notes/rl-1903.xml]: Wrong paste, sorry.

**Comment by florianjacob** [comment]
> if you add an existing printer again, any printer settings you have changed (for the given printer) seem to be reset. (so for example if reconfiguration calls lpadmin again)

@deliciouslytyped that's actually my intention here, as the alternative would be that what's written in the nix config and the actual state of the printer can easily diverge, which I wanted to avoid as far as possible. I tried to improve the documentation in that this isn't a one-time thing, but will actually happen on each reboot or restart of cups. "oneshot" means that the `lpadmin` commands are run like a script on startup of cups. The main identifier is the printer `name` here; so subsequent `lpadmin` calls from the service will modify the existing printer.

**Comment by florianjacob** [comment]
Found the cups tests, rewriting them to use this instead of manual lpadmin commands…

**Comment by deliciouslytyped** [comment]
Hm. Ok, I guess that's a reasonable decision, but I still feel some cognitive dissonance given that the stuff is basically mutable and on a first look doesn't appear touch any of the values in question.
TL;DR: I don't know what to do/thing about it, just a little uneasy.

**Comment by florianjacob** [comment]
> and on a first look doesn't appear touch any of the values in question

@deliciouslytyped Not sure what you're reffering to here, could you explain what do you mean by that?

**Comment by florianjacob** [comment]
@GrahamcOfBorg test printing

**Comment by florianjacob** [comment]
@GrahamcOfBorg test printing

**Comment by florianjacob** [comment]
Added support for socket activation, and added a socket activation version to the printing test to catch such errors.

File: nixos/tests/printing.nix
Comment (Mic92) [inline nixos/tests/printing.nix]: Would the following work instead of this `sleep`?

```
$one->waitUntilSucceeds("lpq | grep -q -E '/active.*root.*$fn/'");
```

File: nixos/tests/printing.nix
Comment (Mic92) [inline nixos/tests/printing.nix]: maybe?

```
$server->waitUntilSucceeds("lpq -a | grep -q -E '$fn'");
```

File: nixos/tests/printing.nix
Comment (Mic92) [inline nixos/tests/printing.nix]: Also here:

```
$server->waitUntilSucceeds("lpq -a | grep -q -E 'no entries'");
```

**Comment by florianjacob** [comment]
@Mic92 replaced all but one sleep with waitUntilSuceeds (no sure what I could match on to find out that service-based cups is actually ready and accepts printing jobs), and also find out about `$machine->name()` with which I could replace the `my name = $machine->succeeds('hostname'); chomp $name;` calls inside the testing function.

**Comment by florianjacob** [comment]
@GrahamcOfBorg test printing

**Comment by bbigras** [comment]
Any progress on this?

**Comment by etu** [comment]
> Any progress on this?

Same question again :smile:

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: I suggest
```nix
{
  noInvalidChars = str: all (c: c != "#" && c != "/") (stringToCharacters str);
}
```

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: I'm 99% sure it's not possible because `builtins.match` is only POSIX regex, which doesn't support lookahead which I think would be needed for that.

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: There's `<command>..</command>` for that, also in `deviceUri` and `model`

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: I assume `lpinfo -v` can be used to figure out these URI's? In any case, in current master, `lpinfo -v` gives me an error:
```
$ lpinfo -v
lpinfo: Bad file descriptor
```

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: Wouldn't hurt readability to move this lonely `default` statement up to its corresponding `mkDefault` :)

File: nixos/modules/hardware/printers.nix
Comment (Infinisil) [inline nixos/modules/hardware/printers.nix]: Can you expand on this? I don't get the problem. Why does "the socket file might still not exist" matter when it's a service?

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/hardware/printers.nix
Comment (etu) [inline nixos/modules/hardware/printers.nix]: The part with `\<printer-name\>` breaks eval for me with the following error:
```
intermediate.xml:3157: parser error : error parsing attribute name
and remove printers with <literal>lpadmin -x \<printer-name\></literal>.
                                                           ^
intermediate.xml:3157: parser error : attributes construct error
and remove printers with <literal>lpadmin -x \<printer-name\></literal>.
                                                           ^
intermediate.xml:3157: parser error : Couldn't find end of Start Tag printer-name line 3157
and remove printers with <literal>lpadmin -x \<printer-name\></literal>.
                                                           ^
unable to parse intermediate.xml
```

I was able to build the manual by removing `\<` and `\>` from the description.

**Review by etu — CHANGES_REQUESTED** [review]
Overall this seems to work, except from my note inline in the code here.

I managed to configure my printer and print a testpage.

**Comment by etu** [comment]
@florianjacob Any progress on this?

Could you rebase on master, fix the conflicts and look into the reported issue I had? It would be great to have this in nixos :)

**Comment by florianjacob** [comment]
Still alive, I'll write again when I found the time to integrate the feedback and resolve the issues.

**Comment by etu** [comment]
@florianjacob I'm a bit eager to try to get this in for 19.09, and the freeze is (probably) in a bit less then a month. Is there something I can help out with to get this in shape? I could probably go through the feedback and push changes later today.

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Makes sense, moved up, asserting you meant `mkOption`. ;)

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Unsure about this, on 19.03 with a running cups, this just prints
```
lpinfo -v
network beh
network http
network ipps
network ipp
network lpd
network https
network socket
network smb
```
don't have master available to test though.

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Thank you!

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Learned how to write a multiline bash command. :+1:

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Tried to elaborate. The gist is: regardless of being socket activated or an always running service, commands like `lpinfo -v` interact with cups through a UNIX socket. With service-based cups, this socket file is not yet created when the cups service gets into “activated” stage, which makes commands fail, while with socket activation, systemd will create the socket file upfront.

File: nixos/modules/hardware/printers.nix
Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Seems like that is not how you would escape pointy brackets in docbook. Well, just removed them, people will figure out it's a placeholder.

**Comment by florianjacob** [comment]
Damn, time for 19.09 already… But rebased, resolved conflicts, adressed more change requests. :)

@etu Thanks for your offer! :+1:  In addition to doing another round of proofreading and testing with the new changes (and old ones like the waitUntilSucceeds changes which vanished somehow but are now recreated),
could you please take a look at my expanded description for https://github.com/NixOS/nixpkgs/pull/55510#discussion_r287607053 , and try to reproduce https://github.com/NixOS/nixpkgs/pull/55510#discussion_r287606970 ?
Also, if you know Perl better than I do, could you try to get rid of that `Machine::retry` thing I don't understand like [suggested here](https://github.com/NixOS/nixpkgs/pull/55510#discussion_r255358099)? That would help alot, thank you!

File: nixos/modules/hardware/printers.nix
Comment (worldofpeace) [reply nixos/modules/hardware/printers.nix]: couldn't you check like?
```
nix run -f channel:nixos-unstable  cups
lpinfo -v
```

Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Did not know you could select another channel like that, thank you. I could confirm that combining the unstable `lpinfo` with stable cups service works.

The problem is the `Bad file descriptor` means that the `lpinfo` client can't communicate with the cups service, which probably means the service has some issues, then. It might just be though that @Infinisil 's cups service was just not running. (And in any way, this issue would be unrelated to this PR.)

File: nixos/modules/hardware/printers.nix
Comment (8573) [reply nixos/modules/hardware/printers.nix]: In any XML-based language, one should be able to use the entity references `&lt;` and `&gt;` to represent `<` and `>`.

Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: Could have guessed that, admittedly. Gauging between using the regular `<option>` syntax for required man page arguments and keeping to the lpadmin man page which does not use anything like that, I switched back to `<option>`.

**Comment by aaronjanse** [comment]
In light of https://github.com/NixOS/nixpkgs/issues/53027, would we want to have paper size as a declarative option?

**Comment by etu** [comment]
> In light of #53027, would we want to have paper size as a declarative option?

That could be a very nice thing, at least for us that aren't in the US.

But at the moment I'm more worried if we get it in at all before the release.

Freeze in 10 days. https://discourse.nixos.org/t/nixos-19-09-feature-freeze/3707

**Comment by florianjacob** [comment]
Just resolved the conflicts, continuing to have both always-running and socket-activated cups in the tests because why not when I already wrote the code for both. But I don't cling to that.

@aaronjanse @etu thank you for reminding me, I already had the code to specify arbitrary ppd options laying around but forgot it was not yet included in this PR. Added those options, handy for working drivers with duplex print and the like, though they obviously don't help against broken drivers like in  #53027.

Calling in @matthewbauer (:+1: for making socket activation the new default) as new cups module maintainer for decision on what to do with this.

**Comment by florianjacob** [comment]
@GrahamcOfBorg test printing

**Review by etu — APPROVED** [review]
Seems to work perfectly fine for me, even with setting page size etc :)

**Review by worldofpeace — APPROVED** [review]
Let's go :smile:  :sparkles:

Thanks so much @florianjacob.

**Comment by florianjacob** [comment]
@etu @worldofpeace  thanks! :)

To anybody lurking here and wanting to try this out, you can use this snippet here on stable 19.03 and then start to configure your printers declaratively right now:

```nix
{ ... }:
let
  declarativePrintersNixpkgsFJ = builtins.fetchGit {
    name = "florianjacob-nixpkgs-declarative-printers";
    url = https://github.com/florianjacob/nixpkgs/;
    ref = "declarative-printers";
    rev = "18a5d23b55d0b0d1bff6bfd971d471d3064011c2";
  };
in {
  imports = [
    "${declarativePrintersNixpkgsFJ}/nixos/modules/hardware/printers.nix"
  ];
}
```

**Comment by etu** [comment]
If nobody comes with any real complaints about this before Friday (2019-09-06), I intend to merge it by then because it works perfectly fine in my tests.

cc @disassembler @Infinisil @Mic92

**Review by infinisil — APPROVED** [review]

**Comment by Infinisil** [comment]
Looking good, let's go :)

File: nixos/modules/hardware/printers.nix
Comment (edolstra) [inline nixos/modules/hardware/printers.nix]: Can we drop the "ensure" part? It seems redundant.

Comment (florianjacob) [reply nixos/modules/hardware/printers.nix]: @edolstra You see a conceptual difference between `ensurePrinters` and `ensureDefaultPrinter` here, and therefore it's redundant for `defaultPrinter` but not for `ensurePrinters`, is that correct? I'm at your service. Maybe you could provide some general guidance on this, though?

I started to use those `ensure` prefixes to mark options that can only be enforced on a best effort basis on each service reload. In this concrete case, both `ensurePrinters` and `ensureDefaultPrinter` are translated into `lpadmin` command. I can't stop the user from modifying printers / change default printer, but the state will be restored. In case of `ensurePrinters`, I deliberately do not try to delete non-listed printers that the user might have added manually. Similar options in NixOS exist though which don't have that prefix.

[There's a discussion going on about that, sparked off by similar mysql / postgresql options](https://github.com/NixOS/nixpkgs/pull/55510)
I explained [my point there](https://github.com/NixOS/nixpkgs/pull/68353#issuecomment-555250666) as well.

**Pull request merged**
